File: onto-agent-server/src/models/datasource.py
import uuid
from datetime import datetime
from typing import Optional
from sqlalchemy import String, Integer, Text, DateTime, func
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.dialects.postgresql import UUID
from src.database import Base


class Datasource(Base):
    __tablename__ = "datasources"

    id: Mapped[str] = mapped_column(
        UUID(as_uuid=False), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    type: Mapped[str] = mapped_column(String(50), nullable=False)
    host: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    port: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    database: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    schema: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    username: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    password: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    ssl_mode: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    description: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(50), default="disconnected")
    table_count: Mapped[int] = mapped_column(Integer, default=0)
    last_sync_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime, server_default=func.now(), onupdate=func.now()
    )

File: onto-agent-server/src/entities/Datasource.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm"

export enum DatasourceType {
  POSTGRESQL = "postgresql",
  MYSQL = "mysql",
  SQLSERVER = "sqlserver",
  ORACLE = "oracle",
}

export enum DatasourceStatus {
  CONNECTED = "connected",
  DISCONNECTED = "disconnected",
  ERROR = "error",
}

@Entity("datasources")
export class Datasource {
  @PrimaryGeneratedColumn("uuid")
  id!: string

  @Column()
  name!: string

  @Column({ type: "varchar", length: 50 })
  type!: DatasourceType

  @Column({ type: "varchar", length: 255, nullable: true })
  host!: string | null

  @Column({ type: "int", nullable: true })
  port!: number | null

  @Column({ type: "varchar", length: 100, nullable: true })
  database!: string | null

  @Column({ type: "varchar", length: 100, nullable: true })
  schema!: string | null

  @Column({ type: "varchar", length: 100, nullable: true })
  username!: string | null

  @Column({ type: "text", nullable: true })
  password!: string | null

  @Column({ type: "varchar", length: 50, nullable: true })
  sslMode!: string | null

  @Column({
    type: "varchar",
    length: 50,
    default: DatasourceStatus.DISCONNECTED,
  })
  status!: DatasourceStatus

  @Column({ type: "int", default: 0 })
  tableCount!: number

  @Column({ type: "timestamp", nullable: true })
  lastSyncAt!: Date | null

  @CreateDateColumn()
  createdAt!: Date

  @UpdateDateColumn()
  updatedAt!: Date
}
